tests/test_json_schema.py

```diff
@@ -2378,15 +2389,28 @@ def __get_pydantic_core_schema__(
             cls, source: Any, handler: Callable[[Any], core_schema.CoreSchema]
         ) -> core_schema.CoreSchema:
             if hasattr(source, '__args__'):
-                param = source.__args__[0]
-                metadata = build_metadata_dict(js_cs_override=handler(Optional[param]))
+                arg = source.__args__[0]
+
+                def js_func(s, h):
+                    # ignore the schema we were given and get a new CoreSchema
+                    s = handler(Optional[arg])
+                    return h(s)
+
                 return core_schema.general_plain_validator_function(
                     Gen,
-                    metadata=metadata,
+                    metadata={'pydantic_js_functions': [js_func]},
                 )
+            else:
+                return handler(source)
 
         @classmethod
-        def __pydantic_modify_json_schema__(cls, field_schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
+        def __get_pydantic_json_schema__(
+            cls, core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
+        ) -> JsonSchemaValue:
+            try:
+                field_schema = handler(core_schema)
+            except PydanticInvalidForJsonSchema:
+                field_schema = {}
```

Comment:
This now fails for the case where `Gen` is not parametrized (as per above it returns an isinstance schema)